zuko/transforms.py

```diff
@@ -629,42 +629,112 @@ class BernsteinTransform(MonotonicTransform):
     bijective = True
     sign = +1
 
-    def __init__(self, theta: Tensor, linear: bool = False, **kwargs):
+    def __init__(self, theta: Tensor, smooth_bounds: bool = False, **kwargs):
         super().__init__(None, phi=(theta,), **kwargs)
 
-        self.theta = self._increasing(theta)
-        self.linear = linear
+        self.theta = self._increasing(theta, smooth_bounds)
 
-        degree = theta.shape[-1]
-        alpha = torch.arange(1, degree + 1, device=theta.device, dtype=theta.dtype)
-        beta = torch.arange(degree, 0, -1, device=theta.device, dtype=theta.dtype)
+        self.order = self.theta.shape[-1] - 1
+        self.dtheta = self.order * (self.theta[..., 1:] - self.theta[..., :-1])
 
-        self.basis = torch.distributions.Beta(alpha, beta)
+        self.basis = self.get_bernstein_basis(self.order, device=theta.device, dtype=theta.dtype)
+        self.dbasis = self.get_bernstein_basis(
+            self.order - 1, device=theta.device, dtype=theta.dtype
+        )
+
+        # save slope on boundaries for interpolation
+        x = torch.tensor([self.eps, 1 - self.eps], device=theta.device, dtype=theta.dtype)
+        rank = self.theta.dim()
+        if rank > 1:
+            # add singleton batch dimensions
+            dims = [...] + [None] * (rank - 1)
+            x = x[dims]
```

Comment:
This should not be necessary as `torch.distributions.Beta.log_prob` broadcasts.

Reply:
Test fail if I remove this. Couldn't make it work without.
What would you suggest?